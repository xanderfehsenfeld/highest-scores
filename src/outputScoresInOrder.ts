const parseLine = (line: string) => {
  const indexOfFirstColon = line.indexOf(":");
  const score = line.substring(0, indexOfFirstColon);
  const jsonString = line.substring(indexOfFirstColon + 1).trim();
  const { id } = JSON.parse(jsonString);

  return {
    id,
    score: parseInt(score),
  };
};

export const outputScoresInOrder = (
  filecontent: string,
  linesToSlice: number
) => {
  const linesInFile = filecontent.split("\n").filter((line) => !!line.trim());
  const selectedLines = linesInFile.slice(0, linesToSlice);
  selectedLines.reverse();
  const scores = selectedLines.map(parseLine);
  scores.sort((a, b) => b.score - a.score);
  return scores;
};
