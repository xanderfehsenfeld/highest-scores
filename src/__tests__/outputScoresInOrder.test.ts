import * as path from "path";
import * as fs from "fs";
import { outputScoresInOrder } from "../outputScoresInOrder";

describe("outputScoresInOrder", () => {
  const fileContentWithoutError = fs.readFileSync(
    path.join(__dirname, "./dataWithoutError.data"),
    "utf-8"
  );
  describe("file length is greater than or equal to selection", () => {
    it("returns selection number of scores ", () => {
      const result = outputScoresInOrder(fileContentWithoutError, 1);
      expect(result).toHaveLength(1);
    });
  });
  describe("file length is less than selection", () => {
    const result = outputScoresInOrder(fileContentWithoutError, 10);
    it("returns the file length number of scores", () => {
      expect(result).toHaveLength(5);
    });
    it("matches snapshot", () => {
      expect(result).toMatchSnapshot();
    });
  });

  it("returns correctly structured JSON", () => {
    const [first] = outputScoresInOrder(fileContentWithoutError, 10);
    expect(first).toHaveProperty("score");
    expect(first).toHaveProperty("id");

    const { score, id } = first;

    expect(typeof score).toBe("number");
    expect(typeof id).toBe("string");
  });

  describe("malformed json", () => {
    const fileContentWithError = fs.readFileSync(
      path.join(__dirname, "./dataWithError.data"),
      "utf-8"
    );
    it("returns selection if error line is excluded", () => {
      const result = outputScoresInOrder(fileContentWithError, 3);
      expect(result).toHaveLength(3);
    });

    it("throws error if error line is included", () => {
      const testWithError = () => outputScoresInOrder(fileContentWithError, 10);
      expect(testWithError).toThrow();
    });
  });

  it("returns empty for empty file", () => {
    const result = outputScoresInOrder(" ", 2);
    expect(result).toHaveLength(0);
  });

  it("returns scores in descending order, highest first", () => {
    const [first, second] = outputScoresInOrder(fileContentWithoutError, 10);
    expect(first.score).toBeGreaterThan(second.score);
  });

  describe("repeated scores", () => {
    const fileContentWithRepeatedScores = fs.readFileSync(
      path.join(__dirname, "./repeatedScores.data"),
      "utf-8"
    );

    const result = outputScoresInOrder(fileContentWithRepeatedScores, 10);
    it("matches snapshot", () => {
      expect(result).toMatchSnapshot();
    });
    it("orders the last occurence first", () => {
      const [first] = result;
      expect(first.id).toEqual("last");
    });
  });
});
