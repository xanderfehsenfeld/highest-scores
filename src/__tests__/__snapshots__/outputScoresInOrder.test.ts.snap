// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputScoresInOrder file length is less than selection matches snapshot 1`] = `
Array [
  Object {
    "id": "085a11e1b82b441184f4a193a3c9a13c",
    "score": 13214012,
  },
  Object {
    "id": "84a0ccfec7d1475b8bfcae1945aea8f0",
    "score": 11446512,
  },
  Object {
    "id": "7ec85fe3aa3c4dd599e23111e7abf5c1",
    "score": 11269569,
  },
  Object {
    "id": "f812d487de244023a6a713e496a8427d",
    "score": 11027069,
  },
  Object {
    "id": "3c867674494e4a7aac9247a9d9a2179c",
    "score": 10622876,
  },
]
`;

exports[`outputScoresInOrder repeated scores matches snapshot 1`] = `
Array [
  Object {
    "id": "last",
    "score": 10622876,
  },
  Object {
    "id": "first",
    "score": 10622876,
  },
]
`;
